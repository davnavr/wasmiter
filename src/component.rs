//! Provides types and functions for parsing the various
//! [components of a WebAssembly module](https://webassembly.github.io/spec/core/syntax/modules.html)
//! from its
//! [sections in the binary format](https://webassembly.github.io/spec/core/binary/modules.html#sections).

mod code_section;
mod datas_component;
mod elems_component;
mod exports_component;
mod funcs_component;
mod function_section;
mod globals_component;
mod imports_component;
mod index_vector;
mod known_section;
mod mems_component;
mod result_type;
mod tables_component;
mod tags_component;
mod type_parser;
mod types_component;

pub use type_parser::{
    block_type, func_type, global_mutability, global_type, limits, mem_type, ref_type, table_type,
    val_type,
};

pub use code_section::{Code, CodeSection, Locals};
pub use datas_component::{DataMode, DatasComponent};
pub use elems_component::{ElementExpressions, ElementInit, ElementMode, ElemsComponent};
pub use exports_component::{Export, ExportKind, ExportsComponent};
pub use funcs_component::{Func, FuncsComponent};
pub use function_section::FunctionSection;
pub use globals_component::GlobalsComponent;
pub use imports_component::{Import, ImportKind, ImportsComponent};
pub use index_vector::IndexVector;
pub use known_section::KnownSection;
pub use mems_component::MemsComponent;
pub use result_type::ResultType;
pub use tables_component::TablesComponent;
pub use tags_component::{parse as tag, Tag, TagsComponent};
pub use types_component::TypesComponent;

/// Parses a
/// [WebAssembly index](https://webassembly.github.io/spec/core/binary/modules.html#indices).
pub fn index<N: crate::index::Index, B: crate::bytes::Bytes>(
    offset: &mut u64,
    bytes: B,
) -> crate::parser::Result<N> {
    use crate::parser::{leb128, ResultExt as _};

    let index = leb128::u32(offset, bytes).context(N::NAME)?;
    Ok(N::try_from(index)?)
}
