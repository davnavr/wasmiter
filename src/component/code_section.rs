use crate::{
    component,
    input::{Input, Window},
    instruction_set::InstructionSequence,
    parser::{self, ResultExt as _, Vector},
};
use core::fmt::{Debug, Formatter};

/// Represents an entry
/// in the
/// [*code section*], also known as a
/// [`func`](https://webassembly.github.io/spec/core/binary/modules.html#code-section).
///
/// To allow reading the code section in parallel and skipping of entries, a [`Code`] stores the
/// size, in bytes, of its contents.
///
/// To read [`Code`] from the [*code section*], see the [`CodeSection`] struct.
///
/// [*code section*]: https://webassembly.github.io/spec/core/binary/modules.html#code-section
#[derive(Clone, Copy)]
pub struct Code<I: Input> {
    index: u32,
    content: Window<I>,
}

impl<I: Input> Code<I> {
    /// The index of this *code section* entry.
    #[inline]
    pub fn index(&self) -> u32 {
        self.index
    }

    /// Gets the binary contents of this *code section* entry.
    #[inline]
    pub fn content(&self) -> &Window<I> {
        &self.content
    }

    /// Reads the contents of this code entry.
    ///
    /// The first closure is given a [`Locals`](component::Locals) used to read the compressed
    /// local variable declarations.
    ///
    /// The second closure is given the output of the first closure, along with an
    /// [`InstructionSequence`] used to read the function *body*.
    pub fn read<Y, Z, E, L, C>(&self, locals_f: L, code_f: C) -> Result<Z, E>
    where
        E: From<parser::Error>,
        L: FnOnce(&mut component::Locals<&mut u64, &Window<I>>) -> Result<Y, E>,
        C: FnOnce(Y, &mut InstructionSequence<&mut u64, &Window<I>>) -> Result<Z, E>,
    {
        let mut offset = self.content.base();
        let mut locals = component::Locals::new(&mut offset, &self.content)?;
        let code_arg = locals_f(&mut locals)?;
        locals.finish()?;

        let mut code = InstructionSequence::new(&mut offset, &self.content);
        let result = code_f(code_arg, &mut code)?;

        let (_, final_offset) = code.finish()?;
        let final_length = *final_offset - self.content.base();
        if final_length != self.content.length() {
            return Err(crate::parser_bad_format_at_offset!(
                "file" @ offset,
                "expected code entry content to have a length of {} bytes, but got {final_length}",
                self.content.length()
            )
            .into());
        }

        Ok(result)
    }
}

impl<I: Input + Clone> Code<&I> {
    /// Clones the underlying [`Bytes`] of this *code section* entry.
    pub fn cloned(&self) -> Code<I> {
        Code {
            index: self.index,
            content: (&self.content).into(),
        }
    }
}

impl<I: Input> Debug for Code<I> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let mut s = f.debug_struct("Func");
        let result = self.read(
            |locals| Ok(s.field("locals", locals)),
            |s, body| {
                s.field("body", body);
                parser::Result::Ok(())
            },
        );

        if let Err(e) = result {
            s.field("error", &e);
        }

        s.finish()
    }
}

/// Represents the
/// [*code section*](https://webassembly.github.io/spec/core/binary/modules.html#code-section),
/// which corresponds to the
/// [**locals** and **body** fields](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-func)
/// of each function in the
/// [**funcs** component](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-func)
/// of a WebAssembly module.
#[derive(Clone, Copy)]
pub struct CodeSection<I: Input> {
    entries: Vector<u64, I>,
}

impl<I: Input> From<Vector<u64, I>> for CodeSection<I> {
    #[inline]
    fn from(entries: Vector<u64, I>) -> Self {
        Self { entries }
    }
}

impl<I: Input> CodeSection<I> {
    /// Uses the given [`Input`] to read the contents of the *code section* of a module, which
    /// begins at the given `offset`.
    #[inline]
    pub fn new(offset: u64, input: I) -> parser::Result<Self> {
        Vector::parse(offset, input)
            .context("at start of code section")
            .map(Self::from)
    }

    /// Gets the expected remaining number of entries in the *code section* that have yet to be
    /// parsed.
    #[inline]
    pub fn remaining_count(&self) -> u32 {
        self.entries.remaining_count()
    }

    /// Parses the next entry in the *code section*.
    pub fn parse(&mut self) -> parser::Result<Option<Code<&I>>> {
        self.entries
            .advance_with_index(|index, offset, bytes| {
                let size = parser::leb128::u64(offset, bytes).context("code entry size")?;
                let content = Window::with_offset_and_length(bytes, *offset, size);

                *offset = offset.checked_add(size).ok_or_else(|| {
                    crate::parser_bad_input!(
                        crate::input::offset_overflowed(*offset),
                        "unable to advance offset to read next code section entry"
                    )
                })?;

                parser::Result::Ok(Code { index, content })
            })
            .transpose()
            .context("within code section")
    }

    #[inline]
    pub(super) fn borrowed(&self) -> CodeSection<&I> {
        self.entries.borrowed().into()
    }
}

impl<I: Clone + Input> Iterator for CodeSection<I> {
    type Item = parser::Result<Code<I>>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.parse() {
            Ok(None) => None,
            Err(e) => Some(Err(e)),
            Ok(Some(code)) => Some(Ok(code.cloned())),
        }
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        self.entries.size_hint()
    }
}

impl<I: Clone + Input> core::iter::FusedIterator for CodeSection<I> {}

impl<I: Input> Debug for CodeSection<I> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.debug_list().entries(self.borrowed()).finish()
    }
}
