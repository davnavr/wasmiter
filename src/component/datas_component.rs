use crate::bytes::{Bytes, Window};
use crate::index::MemIdx;
use crate::instruction_set::InstructionSequence;
use crate::parser::{self, Offset, Result, ResultExt};
use core::fmt::{Debug, Formatter};

/// Specifies the mode of a
/// [data segment](https://webassembly.github.io/spec/core/syntax/modules.html#data-segments).
#[derive(Clone, Copy)]
pub enum DataMode<O: Offset, B: Bytes> {
    /// A **passive** data segment's elements are copied to a memory using the
    /// [`memory.init`](crate::instruction_set::Instruction::MemoryInit) instruction.
    Passive,
    /// An **active** data segment copies elements into the specified memory, starting at the
    /// offset specified by an expression, during
    /// [instantiation](https://webassembly.github.io/spec/core/exec/modules.html#exec-instantiation)
    /// of the module.
    Active(MemIdx, InstructionSequence<O, B>),
}

impl<O: Offset, B: Bytes> DataMode<O, B> {
    fn finish(self) -> Result<()> {
        match self {
            Self::Passive => (),
            Self::Active(_, offset) => {
                offset.finish()?;
            }
        }
        Ok(())
    }
}

impl<O: Offset, B: Bytes> Debug for DataMode<O, B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Passive => f.debug_tuple("Passive").finish(),
            Self::Active(memory, offset) => f
                .debug_struct("Active")
                .field("memory", memory)
                .field("offset", offset)
                .finish(),
        }
    }
}

/// Represents the
/// [**datas** component](https://webassembly.github.io/spec/core/syntax/modules.html#data-segments)
/// of a WebAssembly module, stored in and parsed from the
/// [*data section*](https://webassembly.github.io/spec/core/binary/modules.html#data-section).
#[derive(Clone, Copy)]
pub struct DatasComponent<B: Bytes> {
    count: u32,
    offset: u64,
    bytes: B,
}

impl<B: Bytes> DatasComponent<B> {
    /// Uses the given [`Bytes`] to read the contents of the *data section* of a module, starting
    /// at the specified `offset`.
    pub fn new(mut offset: u64, bytes: B) -> Result<Self> {
        Ok(Self {
            count: parser::leb128::u32(&mut offset, &bytes).context("data section count")?,
            bytes,
            offset,
        })
    }

    #[inline]
    fn parse_inner<Y, Z, M, D>(&mut self, mode_f: M, data_f: D) -> Result<Z>
    where
        M: FnOnce(&mut DataMode<&mut u64, &B>) -> Result<Y>,
        D: FnOnce(Y, Window<&B>) -> Result<Z>,
    {
        let mode_tag =
            parser::leb128::u32(&mut self.offset, &self.bytes).context("data segment mode")?;

        let mut mode = match mode_tag {
            0 => DataMode::Active(
                MemIdx::from(0u8),
                InstructionSequence::new(&mut self.offset, &self.bytes),
            ),
            _ => {
                return Err(crate::parser_bad_format!(
                    "{mode_tag} is not a supported data segment mode"
                ))
            }
        };

        let data_arg = mode_f(&mut mode)?;
        mode.finish()?;

        let data_length =
            parser::leb128::u64(&mut self.offset, &self.bytes).context("data segment length")?;

        let data = Window::new(&self.bytes, self.offset, data_length);
        let result = data_f(data_arg, data)?;

        self.offset = self.offset.checked_add(data_length).ok_or_else(|| {
            crate::parser_bad_format!(
                "expected data segment to have a length of {data_length} bytes, but end of section was unexpectedly reached"
            )
        })?;

        Ok(result)
    }

    /// Parses the next data segment in the section.
    pub fn parse<Y, Z, M, D>(&mut self, mode_f: M, data_f: D) -> Result<Option<Z>>
    where
        M: FnOnce(&mut DataMode<&mut u64, &B>) -> Result<Y>,
        D: FnOnce(Y, Window<&B>) -> Result<Z>,
    {
        if self.count == 0 {
            return Ok(None);
        }

        let result = self.parse_inner(mode_f, data_f);

        if result.is_ok() {
            self.count -= 1;
        } else {
            self.count = 0;
        }

        result.map(Some)
    }

    /// Gets the expected remaining number of entires in the *data section* that have yet to be parsed.
    #[inline]
    pub fn len(&self) -> u32 {
        self.count
    }

    /// Returns a value indicating if the *data section* is empty.
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

impl<B: Bytes> Debug for DatasComponent<B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let mut datas = DatasComponent {
            count: self.count,
            offset: self.offset,
            bytes: &self.bytes,
        };

        struct DataSegment<'a, B: Bytes> {
            mode: DataMode<u64, B>,
            data: Window<&'a B>,
        }

        impl<B: Bytes> Debug for DataSegment<'_, B> {
            fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
                f.debug_struct("DataSegment")
                    .field("mode", &self.mode)
                    .field("data", &self.data)
                    .finish()
            }
        }

        let mut list = f.debug_list();
        loop {
            let result = datas.parse(
                |mode| {
                    Ok(match mode {
                        DataMode::Passive => DataMode::Passive,
                        DataMode::Active(memory, offset) => {
                            DataMode::Active(*memory, offset.map_bytes(|_| &self.bytes))
                        }
                    })
                },
                |mode, data| {
                    list.entry(&Result::Ok(DataSegment { mode, data }));
                    Ok(())
                },
            );

            match result {
                Ok(Some(())) => (),
                Ok(None) => break,
                Err(e) => {
                    list.entry(&Result::<()>::Err(e));
                    break;
                }
            }
        }
        list.finish()
    }
}
