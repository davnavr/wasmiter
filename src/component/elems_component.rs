use crate::{
    bytes::Bytes,
    component,
    index::{self, TableIdx},
    instruction_set::InstructionSequence,
    parser::{self, Offset, Result, ResultExt, Vector},
};
use core::fmt::{Debug, Formatter};

/// Represents a vector of expressions that evaluate to references in an
/// [element segment](https://webassembly.github.io/spec/core/syntax/modules.html#element-segments).
pub struct ElementExpressions<O: Offset, B: Bytes> {
    count: u32,
    offset: O,
    bytes: B,
}

impl<O: Offset, B: Bytes> ElementExpressions<O, B> {
    fn new(mut offset: O, bytes: B) -> Result<Self> {
        Ok(Self {
            count: parser::leb128::u32(offset.offset_mut(), &bytes)
                .context("element segment expression count")?,
            offset,
            bytes,
        })
    }

    fn next_inner<T, F>(&mut self, f: F) -> Result<T>
    where
        F: FnOnce(&mut InstructionSequence<&mut u64, &B>) -> Result<T>,
    {
        let mut offset_cell = self.offset.offset();
        let mut expression = InstructionSequence::new(&mut offset_cell, &self.bytes);
        let result = f(&mut expression)?;
        let (_, final_offset) = expression.finish()?;
        *self.offset.offset_mut() = *final_offset;
        Ok(result)
    }

    /// Parses the next expression.
    pub fn next<T, F>(&mut self, f: F) -> Result<Option<T>>
    where
        F: FnOnce(&mut InstructionSequence<&mut u64, &B>) -> Result<T>,
    {
        if self.count == 0 {
            return Ok(None);
        }

        let result = self.next_inner(f);

        if result.is_ok() {
            self.count -= 1;
        } else {
            self.count = 0;
        }

        result.map(Some)
    }

    fn finish(mut self) -> Result<()> {
        while let Some(Ok(())) = self.next(|_| Ok(())).transpose() {}
        Ok(())
    }
}

impl<O: Offset, B: Bytes> Debug for ElementExpressions<O, B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let mut borrowed = ElementExpressions {
            count: self.count,
            offset: self.offset.offset(),
            bytes: &self.bytes,
        };

        let mut list = f.debug_list();
        loop {
            let result = borrowed.next(|instructions| {
                list.entry(&Result::Ok(instructions));
                Ok(())
            });

            match result {
                Ok(Some(())) => (),
                Ok(None) => break,
                Err(e) => {
                    list.entry(&Result::<()>::Err(e));
                    break;
                }
            }
        }
        list.finish()
    }
}

/// Represents the references within an
/// [element segment](https://webassembly.github.io/spec/core/syntax/modules.html#element-segments).
pub enum ElementInit<O: Offset, B: Bytes> {
    /// A vector of functions to create `funcref` elements from.
    Functions(Vector<O, B, parser::SimpleParse<index::FuncIdx>>),
    /// A vector of expressions that evaluate to references.
    Expressions(Option<crate::types::RefType>, ElementExpressions<O, B>),
}

impl<O: Offset, B: Bytes> ElementInit<O, B> {
    fn finish(self) -> Result<()> {
        match self {
            Self::Functions(functions) => {
                functions.finish()?;
            }
            Self::Expressions(_, expressions) => expressions.finish()?,
        }
        Ok(())
    }
}

impl<O: Offset, B: Bytes> Debug for ElementInit<O, B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Functions(functions) => f.debug_tuple("Functions").field(functions).finish(),
            Self::Expressions(rtype, expressions) => f
                .debug_struct("Expressions")
                .field("type", rtype)
                .field("values", expressions)
                .finish(),
        }
    }
}

/// Specifies a kind of [element segment](https://webassembly.github.io/spec/core/syntax/modules.html#element-segments).
pub enum ElementMode<O: Offset, B: Bytes> {
    /// A **passive** element segment's elements are copied to a table using the
    /// [`table.init`](crate::instruction_set::Instruction::TableInit) instruction.
    Passive,
    /// An **active** element segment copies elements into the specified table, starting at the
    /// expressed offset specified by an expression, during
    /// [instantiation](https://webassembly.github.io/spec/core/exec/modules.html#exec-instantiation)
    /// of the module.
    Active(TableIdx, InstructionSequence<O, B>),
    /// A **declarative** data segment cannot be used at runtime. It can be used as a hint to
    /// indicate that references to the given elements will be used in code later in the module.
    Declarative,
}

impl<O: Offset, B: Bytes> ElementMode<O, B> {
    fn finish(self) -> Result<()> {
        match self {
            Self::Passive | Self::Declarative => (),
            Self::Active(_, instructions) => {
                instructions.finish()?;
            }
        }
        Ok(())
    }
}

impl<O: Offset, B: Bytes> Debug for ElementMode<O, B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Passive => f.debug_tuple("Passive").finish(),
            Self::Declarative => f.debug_tuple("Declarative").finish(),
            Self::Active(table, offset) => f
                .debug_struct("Active")
                .field("table", &table)
                .field("offset", offset)
                .finish(),
        }
    }
}

/// Represents the
/// [**elems** component](https://webassembly.github.io/spec/core/syntax/modules.html#element-segments) of a
/// WebAssembly module, stored in and parsed from the
/// [*element section*](https://webassembly.github.io/spec/core/binary/modules.html#element-section).
#[derive(Clone, Copy)]
pub struct ElemsComponent<B: Bytes> {
    count: u32,
    offset: u64,
    bytes: B,
}

fn elem_kind<B: Bytes>(offset: &mut u64, bytes: B) -> Result<()> {
    match parser::one_byte_exact(offset, bytes).context("elemkind")? {
        0 => Ok(()),
        bad => Err(crate::parser_bad_format!(
            "{bad:#04X} is not a valid elemkind"
        )),
    }
}

impl<B: Bytes> ElemsComponent<B> {
    /// Uses the given [`Bytes`] to read the contents of the *element section* of a module, starting
    /// at the specified `offset`.
    pub fn new(mut offset: u64, bytes: B) -> Result<Self> {
        Ok(Self {
            count: parser::leb128::u32(&mut offset, &bytes).context("element section count")?,
            bytes,
            offset,
        })
    }

    #[inline]
    fn parse_inner<Y, Z, M, I>(&mut self, mode_f: M, init_f: I) -> Result<Z>
    where
        M: FnOnce(&mut ElementMode<&mut u64, &B>) -> Result<Y>,
        I: FnOnce(Y, &mut ElementInit<&mut u64, &B>) -> Result<Z>,
    {
        let start = self.offset;
        let segment_kind =
            parser::leb128::u32(&mut self.offset, &self.bytes).context("element segment mode")?;

        let mut init;
        let init_arg: Y;
        match segment_kind {
            0 => {
                let mut mode = ElementMode::Active(
                    TableIdx::from(0u8),
                    InstructionSequence::new(&mut self.offset, &self.bytes),
                );

                init_arg = mode_f(&mut mode)?;
                mode.finish()?;
                init = ElementInit::Functions(
                    Vector::new(&mut self.offset, &self.bytes, Default::default())
                        .context("function references in active element segment")?,
                );
            }
            1 => {
                let mut mode = ElementMode::Passive;
                elem_kind(&mut self.offset, &self.bytes)?;
                init_arg = mode_f(&mut mode)?;
                mode.finish()?; // Does nothing
                init = ElementInit::Functions(
                    Vector::new(&mut self.offset, &self.bytes, Default::default())
                        .context("function references in passive element segment")?,
                );
            }
            2 => {
                let mut mode = ElementMode::Active(
                    component::index(&mut self.offset, &self.bytes)?,
                    InstructionSequence::new(&mut self.offset, &self.bytes),
                );

                init_arg = mode_f(&mut mode)?;
                mode.finish()?;
                elem_kind(&mut self.offset, &self.bytes)?;
                init = ElementInit::Functions(
                    Vector::new(&mut self.offset, &self.bytes, Default::default())
                        .context("function references in active element segment")?,
                );
            }
            3 => {
                let mut mode = ElementMode::Declarative;
                elem_kind(&mut self.offset, &self.bytes)?;
                init_arg = mode_f(&mut mode)?;
                mode.finish()?; // Does nothing
                init = ElementInit::Functions(
                    Vector::new(&mut self.offset, &self.bytes, Default::default())
                        .context("function references in declarative element segment")?,
                );
            }
            4 => {
                let mut mode = ElementMode::Active(
                    TableIdx::from(0u8),
                    InstructionSequence::new(&mut self.offset, &self.bytes),
                );

                init_arg = mode_f(&mut mode)?;
                mode.finish()?;
                init = ElementInit::Expressions(
                    None,
                    ElementExpressions::new(&mut self.offset, &self.bytes)
                        .context("expressions in active element segment")?,
                );
            }
            5 => {
                let rtype = component::ref_type(&mut self.offset, &self.bytes)?;
                let mut mode = ElementMode::Passive;
                init_arg = mode_f(&mut mode)?;
                mode.finish()?; // Does nothing
                init = ElementInit::Expressions(
                    Some(rtype),
                    ElementExpressions::new(&mut self.offset, &self.bytes)
                        .context("expressions in passive element segment")?,
                );
            }
            6 => {
                let mut mode = ElementMode::Active(
                    component::index(&mut self.offset, &self.bytes)?,
                    InstructionSequence::new(&mut self.offset, &self.bytes),
                );

                init_arg = mode_f(&mut mode)?;
                mode.finish()?;
                let rtype = component::ref_type(&mut self.offset, &self.bytes)?;
                init = ElementInit::Expressions(
                    Some(rtype),
                    ElementExpressions::new(&mut self.offset, &self.bytes)
                        .context("expressions in active element segment")?,
                );
            }
            7 => {
                let rtype = component::ref_type(&mut self.offset, &self.bytes)?;
                let mut mode = ElementMode::Declarative;
                init_arg = mode_f(&mut mode)?;
                mode.finish()?; // Does nothing
                init = ElementInit::Expressions(
                    Some(rtype),
                    ElementExpressions::new(&mut self.offset, &self.bytes)
                        .context("expressions in declarative element segment")?,
                );
            }
            _ => {
                return Err(crate::parser_bad_format_at_offset!(
                    "file" @ start,
                    "{segment_kind} is not a supported element segment mode"
                ))
            }
        }

        let result = init_f(init_arg, &mut init)?;
        init.finish()?;
        Ok(result)
    }

    /// Parses the next element segment in the section.
    pub fn parse<Y, Z, M, I>(&mut self, mode_f: M, init_f: I) -> Result<Option<Z>>
    where
        M: FnOnce(&mut ElementMode<&mut u64, &B>) -> Result<Y>,
        I: FnOnce(Y, &mut ElementInit<&mut u64, &B>) -> Result<Z>,
    {
        if self.count == 0 {
            return Ok(None);
        }

        let result = self.parse_inner(mode_f, init_f);

        if result.is_ok() {
            self.count -= 1;
        } else {
            self.count = 0;
        }

        result.map(Some)
    }

    /// Gets the expected remaining number of entires in the *element section* that have yet to be parsed.
    #[inline]
    pub fn len(&self) -> u32 {
        self.count
    }

    /// Returns a value indicating if the *element section* is empty.
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

impl<B: Bytes> Debug for ElemsComponent<B> {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let bytes = &self.bytes;
        let mut elems = ElemsComponent {
            count: self.count,
            offset: self.offset,
            bytes,
        };

        let mut list = f.debug_list();

        struct Elem<'a, 'b, 'c, 'd, 'e, B: Bytes> {
            mode: ElementMode<u64, &'a B>,
            elements: &'b mut ElementInit<&'c mut u64, &'d &'e B>,
        }

        impl<B: Bytes> Debug for Elem<'_, '_, '_, '_, '_, B> {
            fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
                f.debug_struct("ElementSegment")
                    .field("mode", &self.mode)
                    .field("elements", self.elements)
                    .finish()
            }
        }

        while elems.count > 0 {
            let result = elems.parse(
                |mode| {
                    Ok(match mode {
                        ElementMode::Passive => ElementMode::Passive,
                        ElementMode::Declarative => ElementMode::Declarative,
                        ElementMode::Active(table, offset) => {
                            ElementMode::Active(*table, offset.cloned())
                        }
                    })
                },
                |mode, elements| {
                    list.entry(&Elem { mode, elements });

                    Ok(())
                },
            );

            if let Err(e) = result {
                list.entry(&Result::<()>::Err(e));
            }
        }
        list.finish()
    }
}
