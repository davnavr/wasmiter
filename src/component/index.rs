use crate::parser::{input::Input, Parser, Result, ResultExt};
use core::fmt::Debug;

/// A [WebAssembly index](https://webassembly.github.io/spec/core/binary/modules.html#indices).
pub trait Index: From<u8> + From<u16> + Debug + Eq + core::hash::Hash + Copy + Ord
where
    usize: From<Self>,
    u32: From<Self>,
{
    /// Parses an encoded [`Index`] value.
    fn parse<I: Input>(parser: &mut Parser<I>) -> Result<Self>;
}

macro_rules! indices {
    ($(
        $(#[$meta:meta])*
        struct $name:ident = $descriptor:literal;
    )*) => {$(
        $(#[$meta])*
        #[derive(Clone, Copy, Eq, Hash, PartialEq, PartialOrd, Ord)]
        #[repr(transparent)]
        pub struct $name(u32);

        impl Debug for $name {
            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
                Debug::fmt(&self.0, f)
            }
        }

        impl From<u8> for $name {
            #[inline]
            fn from(index: u8) -> Self {
                Self(u32::from(index))
            }
        }

        impl From<u16> for $name {
            #[inline]
            fn from(index: u16) -> Self {
                Self(u32::from(index))
            }
        }

        impl From<$name> for usize {
            #[inline]
            fn from(index: $name) -> usize {
                index.0 as usize
            }
        }

        impl From<$name> for u32 {
            #[inline]
            fn from(index: $name) -> u32 {
                index.0
            }
        }

        impl Index for $name {
            fn parse<I: Input>(parser: &mut Parser<I>) -> Result<Self> {
                let index = parser.leb128_u32().context($descriptor)?;
                if usize::try_from(index).is_ok() {
                    Ok(Self(index))
                } else {
                    Err(crate::parser_bad_format!("{} {index} is too large", $descriptor))
                }
            }
        }
    )*};
}

indices! {
    /// A [`typeidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-typeidx),
    /// which is an index into the
    /// [*type section*](https://webassembly.github.io/spec/core/binary/modules.html#type-section).
    struct TypeIdx = "type index";
    /// A [`funcidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-funcidx)
    /// refers to an
    /// [imported function](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-importdesc)
    /// or a function defined in the
    /// [*function section*](https://webassembly.github.io/spec/core/binary/modules.html#function-section).
    struct FuncIdx = "function index";
    /// A [`tableidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-tableidx)
    /// refers to an
    /// [imported table](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-importdesc)
    /// or a table defined in the
    /// [*table section*](https://webassembly.github.io/spec/core/binary/modules.html#table-section).
    struct TableIdx = "table index";
    /// A [`memidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-memidx)
    /// refers to an
    /// [imported memory](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-importdesc)
    /// or a memory defined in the
    /// [*memory section*](https://webassembly.github.io/spec/core/binary/modules.html#memory-section).
    struct MemIdx = "memory index";
    /// https://webassembly.github.io/spec/core/binary/modules.html#binary-globalsec
    /// A [`globalidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-globalidx)
    /// refers to an
    /// [imported global](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-importdesc)
    /// or a global defined in the
    /// [*global section*](https://webassembly.github.io/spec/core/binary/modules.html#global-section).
    struct GlobalIdx = "global index";
    /// An [`elemidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-elemidx)
    /// refers to
    /// [element segments](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-elem)
    /// in the
    /// [*element section*](https://webassembly.github.io/spec/core/binary/modules.html#element-section).
    struct ElemIdx = "element index";
    /// A [`dataidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-dataidx)
    /// refers to
    /// [data segments](https://webassembly.github.io/spec/core/syntax/modules.html#syntax-data)
    /// in the
    /// [*data section*](https://webassembly.github.io/spec/core/binary/modules.html#data-section).
    struct DataIdx = "data index";
    /// A [`localidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-localidx)
    /// refers to the parameters and local variables of a function. The types of each local
    /// variable are defined in the
    /// [*function section*](https://webassembly.github.io/spec/core/binary/modules.html#function-section)
    struct LocalIdx = "local variable index";
    /// A [`labelidx`](https://webassembly.github.io/spec/core/binary/modules.html#binary-labelidx)
    /// refers to
    /// [structured control instructions](https://webassembly.github.io/spec/core/syntax/instructions.html#syntax-instr-control)
    /// within the code of a function.
    struct LabelIdx = "label index";
}
