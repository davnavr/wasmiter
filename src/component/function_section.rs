use crate::component::FuncIdx;
use crate::parser::{input::Input, Parser, Result, ResultExt};

/// Represents the
/// [*function section*](https://webassembly.github.io/spec/core/binary/modules.html#function-section).
pub struct FunctionSection<I: Input> {
    count: usize,
    parser: Parser<I>,
}

impl<I: Input> FunctionSection<I> {
    /// Uses the given [`Parser<I>`] to read the contents of the *function section* of a module.
    pub fn new(mut parser: Parser<I>) -> Result<Self> {
        Ok(Self {
            count: parser.leb128_usize().context("function section count")?,
            parser,
        })
    }

    fn try_clone(&self) -> Result<FunctionSection<I::Fork>> {
        Ok(FunctionSection {
            count: self.count,
            parser: self.parser.fork()?,
        })
    }
}

impl<I: Input> core::iter::Iterator for FunctionSection<I> {
    type Item = Result<FuncIdx>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.count == 0 {
            return None;
        }

        let result = self.parser.index().context("index in function section");
        self.count -= 1;
        Some(result)
    }

    #[inline]
    fn count(self) -> usize {
        self.count
    }

    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) {
        (self.count, Some(self.count))
    }
}

// count is correct, since errors are returned if there are too few elements
impl<I: Input> core::iter::ExactSizeIterator for FunctionSection<I> {
    fn len(&self) -> usize {
        self.count
    }
}

impl<I: Input> core::fmt::Debug for FunctionSection<I> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        let mut list = f.debug_list();
        match self.try_clone() {
            Ok(fork) => list.entries(fork),
            Err(failed) => list.entries(core::iter::once(Result::<()>::Err(failed))),
        }
        .finish()
    }
}
