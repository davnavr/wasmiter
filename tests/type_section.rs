use wasmiter::component::TypesComponent;

#[test]
fn basic_type() {
    let bytes = [
        1u8,  // count
        0x60, // func
        1,    // parameter count
        0x7F, // i32
        1,    // result count
        0x7E, // i64
    ];

    insta::assert_display_snapshot!(TypesComponent::new(0, bytes.as_slice()).unwrap());
}
