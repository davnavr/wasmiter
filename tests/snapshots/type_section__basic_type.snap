---
source: tests/type_section.rs
expression: "TypesComponent::new(0, bytes.as_slice()).unwrap()"
---
(type (; 0 ;) (func (param i32) (result i64)))

